File: python-legacy/scripts/lib/fetchers/weixin.py
"""微信公众号抓取器。"""

import re
import requests
from ..interfaces import FetchResult
from ..registry import register_fetcher

HEADERS = {
    "Referer": "https://mp.weixin.qq.com",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
}


@register_fetcher
class WeixinFetcher:
    def can_handle(self, url: str) -> bool:
        return "mp.weixin.qq.com" in url

    def fetch(self, url: str) -> FetchResult:
        resp = requests.get(url, headers=HEADERS, timeout=15)
        resp.raise_for_status()
        html = resp.text

        title = self._extract(html, r'id="activity-name"[^>]*>([\s\S]*?)</')
        author = self._extract(html, r'id="js_author_name"[^>]*>([\s\S]*?)</')
        source_name = self._extract(html, r'id="js_name"[^>]*>([\s\S]*?)</')

        content_match = re.search(r'id="js_content"[^>]*>([\s\S]*?)</div>\s*<script', html)
        if not content_match:
            raise ValueError("Content not found (#js_content)")

        content_html = content_match.group(1)
        has_code_blocks = len(re.findall(r"<pre", content_html)) > 0

        return FetchResult(
            title=title,
            author=author,
            source_name=source_name,
            content=content_html,
            has_code_blocks=has_code_blocks,
        )

    def _extract(self, html: str, pattern: str) -> str:
        match = re.search(pattern, html)
        return re.sub(r"<[^>]+>", "", match.group(1)).strip() if match else ""

File: worker/src/fetchers/weixin.ts
/**
 * WeChat (公众号) article fetcher — direct HTTP with Tavily/Exa fallback.
 */

import type { Fetcher, FetchResult } from "../types/fetcher.js";
import { fetchWithTimeout } from "../lib/http.js";
import { FetchError } from "../lib/errors.js";
import { URL_PATTERN, HEADERS } from "../lib/constants.js";
import type { Logger } from "../lib/log.js";

export const weixinFetcher: Fetcher = {
  canHandle(url: string): boolean {
    return URL_PATTERN.WEIXIN.some((p) => url.includes(p));
  },

  async fetch(url: string, env: Env, log: Logger): Promise<FetchResult> {
    // Try direct fetch first
    const resp = await fetchWithTimeout(url, {
      headers: { Referer: HEADERS.WEIXIN_REFERER, "User-Agent": HEADERS.DEFAULT_UA },
    });
    if (!resp.ok) throw new FetchError(url, `HTTP ${resp.status}`);
    const html = await resp.text();

    // Detect verification page
    if (!html.includes('id="js_content"') || html.includes("环境异常")) {
      log.warn("weixin_blocked", { url });
      // Fallback: Tavily Extract → Exa
      const fallback = (await tryTavily(url, env, log)) ?? (await tryExa(url, env, log));
      if (!fallback) throw new FetchError(url, "Blocked by verification, all fallbacks failed");
      return fallback;
    }

    const title = extractText(html, /id="activity-name"[^>]*>([\s\S]*?)<\//);
    const author = extractText(html, /id="js_author_name"[^>]*>([\s\S]*?)<\//) ||
      extractText(html, /class="rich_media_meta_text"[^>]*>([\s\S]*?)<\//);
    const sourceName = extractText(html, /id="js_name"[^>]*>([\s\S]*?)<\//);

    const contentMatch = html.match(/id="js_content"[^>]*>([\s\S]*?)<\/div>\s*<script/);
    if (!contentMatch) throw new FetchError(url, "Content not found (#js_content)");

    const contentHtml = contentMatch[1] ?? "";
    const hasCodeBlocks = (contentHtml.match(/<pre/g) ?? []).length > 0;

    return { title, author, sourceName, html: contentHtml, hasCodeBlocks };
  },
};

function extractText(html: string, pattern: RegExp): string {
  return (html.match(pattern)?.[1] ?? "").replace(/<[^>]+>/g, "").trim();
}

/** Tavily Extract API — advanced mode for protected pages. */
async function tryTavily(url: string, env: Env, log: Logger): Promise<FetchResult | null> {
  const key = env.TAVILY_API_KEY || (await env.KV.get("tavily_api_key"));
  if (!key) return null;
  try {
    log.info("tavily_attempt", { url });
    const resp = await fetchWithTimeout("https://api.tavily.com/extract", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ api_key: key, urls: [url], extract_depth: "advanced" }),
      timeoutMs: 30_000,
    });
    if (!resp.ok) { log.warn("tavily_failed", { status: String(resp.status) }); return null; }
    const data = await resp.json<{ results?: { raw_content?: string; title?: string }[] }>();
    const content = data.results?.[0]?.raw_content;
    if (!content || content.length < 100) { log.warn("tavily_empty"); return null; }
    const title = data.results?.[0]?.title || extractTitleFromMd(content);
    log.info("tavily_success", { title, chars: String(content.length) });
    return { title, author: "", sourceName: "WeChat", html: content, hasCodeBlocks: false, isMarkdown: true };
  } catch (e: unknown) {
    log.error("tavily_error", { error: e instanceof Error ? e.message : "unknown" });
    return null;
  }
}

/** Exa get-contents API fallback. */
async function tryExa(url: string, env: Env, log: Logger): Promise<FetchResult | null> {
  const key = env.EXA_API_KEY || (await env.KV.get("exa_api_key"));
  if (!key) return null;
  try {
    log.info("exa_attempt", { url });
    const resp = await fetchWithTimeout("https://api.exa.ai/contents", {
      method: "POST",
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
      body: JSON.stringify({ urls: [url], text: true }),
      timeoutMs: 30_000,
    });
    if (!resp.ok) { log.warn("exa_failed", { status: String(resp.status) }); return null; }
    const data = await resp.json<{ results?: { text?: string; title?: string; author?: string }[] }>();
    const item = data.results?.[0];
    if (!item?.text || item.text.length < 100) { log.warn("exa_empty"); return null; }
    log.info("exa_success", { title: item.title || "", chars: String(item.text.length) });
    return { title: item.title || "", author: item.author || "", sourceName: "WeChat", html: item.text, hasCodeBlocks: false, isMarkdown: true };
  } catch (e: unknown) {
    log.error("exa_error", { error: e instanceof Error ? e.message : "unknown" });
    return null;
  }
}

function extractTitleFromMd(md: string): string {
  const m = md.match(/^#\s+(.+)/m);
  return m?.[1] ?? "";
}
